src/wp-includes/class-wp-block-metadata-registry.php

```diff
@@ -0,0 +1,227 @@
+<?php
+/**
+ * Block Metadata Registry
+ *
+ * @package WordPress
+ * @subpackage Blocks
+ * @since 6.X.0
+ */
+
+/**
+ * Class used for managing block metadata collections.
+ *
+ * The WP_Block_Metadata_Registry allows plugins to register metadata for large
+ * collections of blocks (e.g., 50-100+) using a single PHP file. This approach
+ * reduces the need to read and decode multiple `block.json` files, enhancing
+ * performance through opcode caching.
+ *
+ * @since 6.X.0
+ */
+class WP_Block_Metadata_Registry {
+
+	/**
+	 * Container for storing block metadata collections.
+	 *
+	 * Each entry maps a base path to its corresponding metadata and callback.
+	 *
+	 * @since 6.X.0
+	 * @var array<string, array<string, mixed>>
+	 */
+	private static $collections = array();
+
+	/**
+	 * Caches the last matched collection path for performance optimization.
+	 *
+	 * @since 6.X.0
+	 * @var string|null
+	 */
+	private static $last_matched_collection = null;
+
+	/**
+	 * Registers a block metadata collection.
+	 *
+	 * This method allows registering a collection of block metadata from a single
+	 * manifest file, improving performance for large sets of blocks.
+	 *
+	 * The manifest file should be a PHP file that returns an associative array, where
+	 * the keys are the block identifiers and the values are the corresponding block
+	 * metadata arrays. The block identifiers are determined by the default identifier
+	 * callback, which uses the parent directory name for 'block.json' files and the
+	 * directory name itself for directories.
```

Comment:
Since you can't provide a custom callback as of now, a bit too much detail to mention the "default identifier callback".
```suggestion
	 * metadata arrays. The block identifiers must match the parent directory name for
	 * the respective `block.json` file.
```